src/dbus/xdg_deskop_portal_impl_access.rs

```diff
@@ -0,0 +1,150 @@
+use std::ops::Not;
+
+use zbus::dbus_interface;
+
+use super::Start;
+
+pub struct AccessPortalImpl {
+    to_niri: calloop::channel::Sender<AccessDialogRequest>,
+}
+
+#[dbus_interface(name = "org.freedesktop.impl.portal.Access")]
+impl AccessPortalImpl {
+    /// AccessDialog method
+    #[allow(clippy::too_many_arguments)]
+    async fn access_dialog(
+        &self,
+        _handle: zbus::zvariant::ObjectPath<'_>,
+        app_id: &str,
+        parent_window: &str,
+        title: &str,
+        subtitle: &str,
+        body: &str,
+        options: std::collections::HashMap<&str, zbus::zvariant::Value<'_>>,
+    ) -> zbus::fdo::Result<(
+        u32,
+        std::collections::HashMap<String, zbus::zvariant::OwnedValue>,
+    )> {
+        let (response_channel_sender, response_channel_receiver) = async_channel::bounded(1);
+
+        let options = AccessDialogOptions::from_options(options);
+        let request = AccessDialogRequest {
+            app_id: app_id.to_string(),
+            parent_window: parent_window
+                .is_empty()
+                .not()
+                .then(|| parent_window.to_string()),
+            title: title.to_string(),
+            subtitle: subtitle.to_string(),
+            body: body.is_empty().not().then(|| body.to_string()),
+            options,
+
+            response_channel_sender,
+        };
+
+        if let Err(err) = self.to_niri.send(request) {
+            tracing::warn!(?err, "failed to send access dialog request");
+            return Ok((2, Default::default()));
+        };
+
+        let result = std::collections::HashMap::<String, zbus::zvariant::OwnedValue>::new();
+        let response = match response_channel_receiver.recv().await {
+            Ok(AccessDialogResponse::Grant) => {
+                // FIXME: Add selected choices to the result
+                0
+            }
+            Ok(AccessDialogResponse::Deny) => 1,
+            Err(err) => {
+                tracing::warn!(?err, "failed to receive response for access dialog request");
+                return Ok((2, Default::default()));
+            }
+        };
+
+        return Ok((response, result));
+    }
+}
+
+impl AccessPortalImpl {
+    pub fn new(to_niri: calloop::channel::Sender<AccessDialogRequest>) -> Self {
+        Self { to_niri }
+    }
+}
+
+impl Start for AccessPortalImpl {
+    fn start(self) -> anyhow::Result<zbus::blocking::Connection> {
+        let conn = zbus::blocking::ConnectionBuilder::session()?
+            .name("org.freedesktop.impl.portal.desktop.niri")?
+            .serve_at("/org/freedesktop/portal/desktop", self)?
+            .build()?;
+        Ok(conn)
+    }
+}
+
+pub struct AccessDialogRequest {
```

Comment:
Could we maybe move these types into `src/ui/access_dialog.rs`? And don't gate that file under dbus. To avoid dbus-gating lots of statements in random places in the code.

Reply:
moved to `ui/access_dialog` and removed the gates